### ISSUE TITLE
Haversine distance for the metrics modules

### ISSUE THREAD

**Issue by Danisodu** [rtavenar/tslearn#162]
It would be nice to have the option to choose other distances, for instance haversine, , to be used in the metrics module.

**Comment by GillesVandewiele** [comment]
This would indeed be a nice suggestion, but for the time being, you can use `https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.FunctionTransformer.html` in combination with an sklearn Pipeline with KNN (or some clustering algorithm).

e.g.:

```
pipe = Pipeline([
	(
		'norm', 
		TimeSeriesScalerMeanVariance()
	),
	(
		'transform', 
		SymbolicAggregateApproximation(n_segments=16, alphabet_size_avg=10)
	),
	(
		'distance', 
		FunctionTransformer(calc_distances, validate=False, pass_y=False)
	)
])

all_X = numpy.vstack((X_train, X_test))
distances = pipe.transform(all_X)

# We only need the distances to the timeseries from the training set.
# Both for the test and the training set.
X_train_dist = distances[:len(X_train), :len(X_train)]
X_test_dist = distances[len(X_train):, :len(X_train)]

knn = KNeighborsClassifier(n_neighbors=1, metric='precomputed')
knn.fit(X_train_dist, y_train)
predictions = knn.predict(X_test_dist)
print('Accuracy score on test set with KNN on SAX using MINDIST')
print(confusion_matrix(y_test, predictions))
print(accuracy_score(y_test, predictions))
```

**Comment by Danisodu** [comment]
I wanted this to explore the metrics module, not clustering

**Comment by GillesVandewiele** [comment]
Well what do you want to do with the metrics module? Do you want a pairwise distance matrix? If so, `distances` from my code snippet will be just that.

**Comment by GillesVandewiele** [comment]
Ok, so the goal would be to allow to define the metric you are using in order to calculate the pairwise distance matrix in which DTW looks for the optimal path.

In other words, instead of using always using Euclidean distances as in the matrix below, we could allow a configurable metric.

![image](https://user-images.githubusercontent.com/5750603/78170088-b6b9ed80-7452-11ea-9259-d807be3f567e.png)

**Comment by Danisodu** [comment]
The traj-dist library (https://github.com/bguillouet/traj-dist) offers the possibility to use either euclidean or spherical distance for different types of similarity measures. But that's just it, it doesn't offer dtw_path and many others functions that tslearn does. Also, the DTW values with euclidean distance from traj-dist differs from the ones with tslearn, don't know why ¯\_(ツ)_/¯

**Comment by rtavenar** [comment]
@Danisodu 

Regarding the differences between implementations, we detail in the docs [how we compute DTW](https://tslearn.readthedocs.io/en/latest/dtw.html#optimization-problem).

Then, having something like `dtw_path` but for another inner cost metric would indeed make sense. What would you think of a function that would have the following prototype?

```python
def dtw_path_from_inner_cost(s1, s2, inner_cost_matrix, global_constraint=None, 
                             sakoe_chiba_radius=None, itakura_max_slope=None)
```

This would not be very difficult to implement, but I'm afraid I would not have time to do so in the near future. I could definitely help anyone on that, though.

**Comment by Danisodu** [comment]
I don't think there's the need to create a new function, just change the existing ones to have a new signature where the distance type (euclidean, haversine, or else) can be provided. 
For now, I will change the tslearn code directly.

**Comment by rtavenar** [comment]
Well, in terms of library design, I think I would prefer to have a different function for this use-case so that the docs can be more concise for the general case (Euclidean) and advanced users can read through another function that meets their needs.

**Comment by Danisodu** [comment]
Ok ok. Of course it would be nice to extend this to all functions of dtw, but having it for both dtw() and dtw_path() would suffice already

**Comment by rtavenar** [comment]
@Danisodu 

A fix has been implemented in #219 . It will be available in tslearn v0.4.
If you want to use it before v0.4 is out, you should first install the tslearn from @rfayat repo (who implemented the feature):

```bash
python -m pip install https://github.com/rfayat/tslearn/archive/dtw_path_inner_cost.zip
```

and here is an excerpt from the docs:

> The wrapping can be done by passing a string indicating the metric to pass to scikit-learn pairwise_distances:
> 
> ```python
> >>> dtw_path_from_metric(s1, s2, metric="sqeuclidean")
> [(0, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)], 1.1177581530156733
> ```
